compiler/aarch64/codegen/OMRTreeEvaluator.cpp

```diff
@@ -5525,9 +5525,37 @@ TR::Register *commonLoadEvaluator(TR::Node *node, TR::InstOpCode::Mnemonic op, i
     TR::MemoryReference *tempMR = TR::MemoryReference::createWithRootLoadOrStore(cg, node);
     tempMR->validateImmediateOffsetAlignment(node, size, cg);
 
-    generateTrg1MemInstruction(cg, op, node, targetReg, tempMR);
+    if (needSync && tempMR->getUnresolvedSnippet() == NULL) {
+        TR_ASSERT_FATAL_WITH_NODE(node, (size == 1) || (size == 2) || (size == 4) || (size == 8),
+            "operand size for ldar must be 1, 2, 4, or 8 bytes");
+        int numberOfBytesLog2 = trailingZeroes(size);
 
-    if (needSync) {
+        static const TR::InstOpCode::Mnemonic ldarOpCodes[]
+            = { TR::InstOpCode::ldarb, TR::InstOpCode::ldarh, TR::InstOpCode::ldarw, TR::InstOpCode::ldarx };
+
+        tempMR->simplify(node, cg);
+
+        TR::Register *targetGPR = targetReg;
+        if (targetReg->getKind() != TR_GPR) {
+            // ldar only operates on GPRs, so we need to load into a GPR then shuffle into targetReg
+            targetGPR = cg->allocateRegister();
+        }
+
+        generateTrg1MemInstruction(cg, ldarOpCodes[numberOfBytesLog2], node, targetGPR, tempMR);
+
+        if (targetReg->getKind() != TR_GPR) {
+            generateTrg1Src1Instruction(cg,
+                targetReg->isSinglePrecision() ? TR::InstOpCode::fmov_wtos : TR::InstOpCode::fmov_xtod, node, targetGPR,
+                targetReg);
+            cg->stopUsingRegister(targetGPR);
+        } else if (size < 4) {
+            generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::sbfmx, node, targetReg, targetReg, size * 8 - 1);
```

Comment:
There are calls to `commonLoadEvaluator()` that don't want sign-extension of byte and half-word data:
https://github.com/eclipse-omr/omr/blob/aa214b485f6446741a1caf5855ed0d38bc6c5b89/compiler/aarch64/codegen/UnaryEvaluator.cpp#L530-L562